src/components/icon/icon-registry.ts

```diff
@@ -0,0 +1,239 @@
+      const config = this._iconConfigs.get(namespace).get(iconName);
+      if (config.svgElement) {
+        // We already have the SVG element for this icon, return a copy.
+        return Observable.of(config.svgElement.cloneNode(true));
+      } else {
+        // Fetch the icon from the config's URL, cache it, and return a copy.
+        return this._loadIconFromConfig(config)
+          .do((svg: SVGElement) => config.svgElement = svg)
+          .map((svg: SVGElement) => svg.cloneNode(true));
+      }
+    }
+    // See if we have any icon sets registered for the set name.
+    const iconSetConfigs = this._iconSetConfigs.get(namespace);
+    if (iconSetConfigs) {
+      // For all the icon set SVG elements we've fetched, see if any contain an icon with the
+      // requested name.
+      const namedIcon = this._extractIconWithNameFromAnySet(iconName, iconSetConfigs);
+      if (namedIcon) {
+        // We could cache namedSvg in _iconConfigs, but since we have to make a copy every
+        // time anyway, there's probably not much advantage compared to just always extracting
+        // it from the icon set.
+        return Observable.of(namedIcon);
+      }
+      // Not found in any cached icon sets. If there are icon sets with URLs that we haven't
+      // fetched, fetch them now and look for iconName in the results.
+      const iconSetFetchRequests = <[Observable<SVGElement>]>[];
+      iconSetConfigs.forEach((setConfig) => {
+        if (!setConfig.svgElement) {
+          iconSetFetchRequests.push(
+              this._loadIconSetFromConfig(setConfig)
+                  .catch((err: any, source: any, caught: any): Observable<SVGElement> => {
+                    // Swallow errors fetching individual URLs so the combined Observable won't
+                    // necessarily fail.
+                    console.log(`Loading icon set URL: ${setConfig.url} failed with error: ${err}`);
+                    return Observable.of(null);
+                  })
+                  .do((svg: SVGElement) => {
+                    // Cache SVG element.
+                    if (svg) {
+                      setConfig.svgElement = svg;
+                    }
+                  })
+          );
+        }
+      });
+      // Fetch all the icon set URLs. When the requests complete, every IconSet should have a
+      // cached SVG element (unless the request failed), and we can check again for the icon.
+      return Observable.forkJoin(iconSetFetchRequests)
+        .map((ignoredResults: any) => {
+          const foundIcon = this._extractIconWithNameFromAnySet(iconName, iconSetConfigs);
+          if (!foundIcon) {
+            throw Error(`Failed to find icon name: ${iconName} in namespace: ${namespace}`);
+          }
+          return foundIcon;
+        });
+    }
+    return Observable.throw(Error(`Unknown icon name: ${iconName} in namespace: ${namespace}`));
```

Comment:
We've been using custom error classes to make testing easier (see the `portal/` code for an example)